ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java

```diff
@@ -528,12 +549,42 @@ RouterException getChunkException() {
      *                                    created as part of this poll operation.
      */
     void poll(RequestRegistrationCallback<GetOperation> requestRegistrationCallback) {
-      //First, check if any of the existing requests have timed out.
-      cleanupExpiredInFlightRequests();
-      checkAndMaybeComplete();
-      if (!isComplete()) {
-        fetchRequests(requestRegistrationCallback);
+      if (!mayBeProcessDecryptCallbackAndCompleteOperation()) {
+        //First, check if any of the existing requests have timed out.
+        cleanupExpiredInFlightRequests();
+        checkAndMaybeComplete();
+        if (!isComplete() && !onDecryptMode()) {
+          fetchRequests(requestRegistrationCallback);
+        }
+      }
+    }
+
+    /**
+     * May be process decrypt callback and complete the operation if applicable. This is a no-op for non-encrypted blobs
+     * @return {@code true} if the operation is complete, {@code false} otherwise
+     */
+    boolean mayBeProcessDecryptCallbackAndCompleteOperation() {
+      if (onDecryptMode() && decryptCallbackResultAvailable) {
+        logger.trace("Processing decrypt callback stored result for data chunk {}", chunkBlobId);
+        if (decryptCallbackException == null && (getChunkException() == null || (getChunkException() != null
+            && !retainChunkExceptionOnSuccess))) {
+          chunkIndexToBuffer.put(chunkIndex, filterChunkToRange(decryptCallbackResult.getDecryptedBlobContent()));
+          numChunksRetrieved++;
+          logger.trace("Decrypt result successfully updated for data chunk {}", chunkBlobId);
+        } else {
+          logger.error("Setting operation exception as decryption callback invoked with exception {} for data chunk {}",
+              decryptCallbackException.getCause(), chunkBlobId);
+          setOperationException(
+              new RouterException("Exception thrown on decrypting the content for data chunk " + chunkBlobId,
+                  RouterErrorCode.UnexpectedInternalError));
+        }
+        logger.trace("Marking blob content available to process for data chunk {}", chunkBlobId);
+        blobContentAvailableToProcess = true;
+        checkAndMaybeComplete();
```

Comment:
I also don't think that this method should complete the operation, because the Chunk-level poll method will take care of that, right?

Reply:
Not really. usually the chunk is completed in handleResponse(). In handleResponse() after processing we do two things.
checkAndMaybeComplete();
    if (getChunk.isComplete()) {
      onChunkOperationComplete(getChunk);
    }

and hence I had to do it this way.